src/backend/distributed/utils/connection_management.c

```diff
@@ -0,0 +1,687 @@
+	}
+
+	key.port = port;
+	if (user)
+	{
+		strlcpy(key.user, user, NAMEDATALEN);
+	}
+	else
+	{
+		strlcpy(key.user, CurrentUserName(), NAMEDATALEN);
+	}
+	if (database)
+	{
+		strlcpy(key.database, database, NAMEDATALEN);
+	}
+	else
+	{
+		strlcpy(key.database, get_database_name(MyDatabaseId), NAMEDATALEN);
+	}
+
+	if (CurrentCoordinatedTransactionState == COORD_TRANS_NONE)
+	{
+		CurrentCoordinatedTransactionState = COORD_TRANS_IDLE;
+	}
+
+	/*
+	 * Lookup relevant hash entry. We always enter. If only a cached
+	 * connection is desired, and there's none, we'll simply leave the
+	 * connection list empty.
+	 */
+
+	entry = hash_search(ConnectionHash, &key, HASH_ENTER, &found);
+	if (!found)
+	{
+		entry->connections = MemoryContextAlloc(ConnectionContext,
+												sizeof(dlist_head));
+		dlist_init(entry->connections);
+	}
+
+	/* if desired, check whether there's a usable connection */
+	if (flags & CACHED_CONNECTION)
+	{
+		/* check connection cache for a connection that's not already in use */
+		connection = FindAvailableConnection(entry->connections, flags);
+		if (connection)
+			return connection;
+	}
+
+	/* no connection available, done if a new connection isn't desirable */
+	if (!(flags & NEW_CONNECTION))
+	{
+		return NULL;
+	}
+
+	/*
+	 * Check whether we're right now allowed to open new connections.
+	 *
+	 * FIXME: This should be removed soon, once all connections go through
+	 * this API.
+	 */
+	if (XactModificationLevel > XACT_MODIFICATION_DATA)
+	{
+		ereport(ERROR, (errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+						errmsg("cannot open new connections after the first modification "
+							   "command within a transaction")));
+	}
+
+	/*
+	 * Either no caching desired, or no pre-established, non-claimed,
+	 * connection present. Initiate connection establishment.
+	 */
+	connection = StartConnectionEstablishment(&key);
+
+	oldContext = MemoryContextSwitchTo(ConnectionContext);
+	dlist_push_tail(entry->connections, &connection->node);
+
+	MemoryContextSwitchTo(oldContext);
+
+	if (flags & SESSION_LIFESPAN)
+	{
+		connection->sessionLifespan = true;
+	}
+
+	connection->activeInTransaction = true;
+
+	return connection;
+}
+
+
+/* StartNodeUserDatabaseConnection() helper */
+static MultiConnection *
+FindAvailableConnection(dlist_head *connections, uint32 flags)
+{
+	dlist_iter iter;
+
+	dlist_foreach(iter, connections)
+	{
+		MultiConnection *connection = dlist_container(MultiConnection, node, iter.cur);
+
+		/* don't return claimed connections */
+		if (connection->claimedExclusively)
+		{
+			continue;
+		}
+
+		/*
+		 * If we're not allowed to open new connections right now, and the
+		 * current connection hasn't yet been used in this transaction, we
+		 * can't use it.
+		 */
+		if (!connection->activeInTransaction &&
+			XactModificationLevel > XACT_MODIFICATION_DATA)
+		{
+			continue;
+		}
+
+		if (flags & SESSION_LIFESPAN)
+		{
+			connection->sessionLifespan = true;
+		}
+		connection->activeInTransaction = true;
```

Comment:
Consider the scenario where you have connections open to some nodes, but you haven't used them in a transaction. If we're e.g. doing router executor type transactions, we're not allowed to open new connections after the beginning of a transactions. So we have to keep track whether connections already have been used in a transaction or not.

I agree that this would be better placed in the caller.